pkg/webhook/resources/upgrade/validator_test.go

```
@@ -0,0 +1,387 @@
+package upgrade
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	"github.com/harvester/harvester/pkg/util"
+)
+
+const (
+	node1    = "node1"
+	node2    = "node2"
+	machine1 = "machine1"
+	machine2 = "machine2"
+)
+
+func Test_isNodeMachineMatching(t *testing.T) {
+	tests := []struct {
+		name        string
+		nodes       []*corev1.Node
+		machines    []*clusterv1.Machine
+		expectError bool
+		errorKey    string
+	}{
+		{
+			name: "count mismatch",
```

Good tests :+1: 

Can you add a tests for the case when there is a machine but there is no node?

---

It is checked on this line https://github.com/harvester/harvester/pull/8222/files#diff-b31f14ddefb333104c8045fe30364dd59957a3be3cccc8720dc5ea517665782eR387, and out of the none-fake-client related codes; 

ideally, we have a full test code which has all kinds of fake clients and can cover all cases

in this PR, it goes a short path, just test `isNodeMachineMatching`, and `isNodeMachineMatching` has the precondition that `node` has at least 1 node

---

I think unit-testing `isNodeMachingMatching` on its own without fake clients is fine.

`isNodeMachineMatching` should return an error when there are more machines than nodes, since it checks that the count is equal in line 400. But the tests only cover the case where there are more nodes than machines.
The tests should make sure that any future modification to the code keeps the behavior that `isNodeMachineMatching` returns an error if the lengths of the two lists aren't equal, regardless which one is longer. Therefore both cases should be tested.

I'd move the `== 0` check from line 387 into `isNodeMachineMatching` and add another check testing `if len(machines) == 0`.
This way, we have a nice separation of concerns between the two functions. `checkNodeMachineMatching` is concerned with collecting the necessary data.
`isNodeMachineMatching` is only concerned with analyzing the two lists of objects for inconsistencies and can be unit tested without the need for fake clients and such. Since it's the only place that implements any actual logic for the validator, it makes testing the correct behavior of the validator in all cases fairly straightforward.

---

It is a good idea to make `isNodeMachineMatching` even independent. I have updated the code, please take a new look, thanks.